Filter biobliography results by publication date

remo87 opened the pull request:
Solves issue https://github.com/opentargets/issues/issues/2578

remo87 committed:
add date and comparer arguments and filter lit data

remo87 committed:
format files using scalafmt

JarrodBaker commented on app/models/Backend.scala:
```diff
@@ -1,5 +1,7 @@
 package models
 
+import akka.http.scaladsl.model.DateTime
+import cats.kernel.Comparison.GreaterThan
```
These imports are unused and can be removed.

remo87 replied:
The imports have been removed from the file

JarrodBaker commented on app/models/Backend.scala:
```diff
@@ -783,6 +808,29 @@ class Backend @Inject() (implicit
     }
   }
 
+  def filterLiteratureByDate(
```
I think this signature can be shorter. `date` and `dateComparator` are both `Option`, but it only makes sense to have one if both fields are of type `Some`. We can wrap an tuple in an `Option` so that we can do a single match on whether we're filtering by date or not:
```
  def filterLiteratureByDate(pub: Publication,
                             filter: Option[(Long, ComparatorEnum.Comparator)]): Boolean = {
    filter match {
      case Some(dataAndComparator) => ???
      case None        => true
    }
```

remo87 replied:
Created a tuple to handle the value and the existence of the values is checked in the invoker method

JarrodBaker commented on app/models/Backend.scala:
```diff
@@ -783,6 +808,29 @@ class Backend @Inject() (implicit
+      if(date.isEmpty) true
+      else {
+        val format = new java.text.SimpleDateFormat("yyyy-MM-dd")
+        val d1: Long = format.parse(pub.date).getTime
```
The filter here doesn't work I don't think. `getTime` returns the time in milliseconds since 1 Jan 1970 so we're going to have very large numbers. Date as provided through by the filter is going to be a year, say 2020. Unless we convert that into Millis first we will get the wrong results. I might be wrong if you are planning on having the FE pass in the date to the API in Millis?

I've checked in the DB and all the records start with YYYY, so you can always do something like `pub.date.take(4)` and use that to compare with the `date` provided. For security you can add in an `assert` statement checking that we extracted a valid year.

remo87 replied:
Add year and month to the query arguments

JarrodBaker commented on app/models/gql/Arguments.scala:
```diff
@@ -63,6 +64,27 @@ object Arguments {
     "Use disease ontology to capture evidences from all descendants to build associations"
   )
 
+  val ComparerEnum = EnumType(
```
You can use a `deriveEnumType` macro here if you wanted to.
See: https://sangria-graphql.github.io/learn/#enumtype-derivation

remo87 replied:
Changed to derived enum

JarrodBaker commented on app/models/entities/Enums.scala:
```diff
@@ -0,0 +1,7 @@
+package models.entities
+
```
This is fine how it is. Another way you'll often see this done in Scala is using a sealed trait. It would look something like:
```
sealed trait Comparison
case object LessThan extends Comparison
case object GreaterThan extends Comparison
```
The `sealed` keyword tells the compiler that all classes that extend the trait are defined in that file. It allows you to match exhaustively so you can have something like:
```
def fun(c: Comparator): Boolean = c match {
  case GreaterThan => ???
  case LessThan => ???
}
```
Without having to include a `case _ => ???` to catch any remaining cases, since the compiler can infer that all valid cases are represented.

As I said, there is no need to change this, I just wanted you to be aware of how you might see the Enum idea implemented elsewhere in the project.

JarrodBaker commented on app/models/gql/Arguments.scala:
```diff
@@ -1,8 +1,9 @@
 package models.gql
 
+import akka.http.scaladsl.model.DateTime
```
Unused import can be removed

remo87 replied:
the import has been removed

remo87 committed:
remove unused imports

remo87 committed:
re format files using scalafmtAll
add year and month to publication query
add year and month to arguments
re factoring filter

remo87 committed:
change EnumType to deriveEnum

JarrodBaker reviewed (dismissed):
As discussed, please update the publication count to use the filtered value and then feel free to merge.

remo87 committed:
update compareDates signature
update total
update year check
